### ISSUE TITLE
Recognize hex notation (0xff0000) as a number when exporting scratch-js

### ISSUE THREAD

**Issue by PullJosh** [PullJosh/sb-edit#9]
See https://github.com/PullJosh/scratch-js/commit/a2ff1a55f5f8d8b57bd35ece4ff77bdc8437650b#commitcomment-36606861

![image](https://user-images.githubusercontent.com/5458180/71564140-ee875280-2a69-11ea-9742-a34c4aa56966.png)
In the situation above, the compiled scratch-js code should be
```js
this.vars.myVariable = 0x7f7f7f; // good
```
rather than
```js
this.vars.myVariable = "0x7f7f7f"; // bad
```

**Comment by towerofnix** [comment]
This was approached in #41 before but I don't think generating an AST rather than direct syntax is a direction anyone here is currently interested in exploring. I think the overall Leopard generation infrastructure is pretty sizeably improved since last time, so I'm self-assigning to take a look at this again.

**Comment by ZXMushroom63** [comment]
Fairly sure that storing the variable as a string is identical to how Scratch would do this. It only recognises the hex as a number in the set color block.

Changing it to a number would break the letter of operator, and probably a few more:
```
var x = 0xff;
x.toString(); //255, so "letter 1 of x" would return '2' instead of '0'
```

**Comment by towerofnix** [comment]
Good catch. It sounds like the fix is basically to wrap blocks in number-shaped inputs with `this.toNumber(...)` more aggressively — namely any time we can't be certain that the contents are already a JavaScript number primitive, because `'24'` and `'0x18'` are both treated as `24` when Scratch is treating values as numbers. (While JavaScript will nicely typecast `'24'` -> `24` in math ops [usually], it doesn't handle `'0x'` prefixes in strings.)

For color inputs in particular we can handle casting in Leopard's `Color.num()` function, i.e. have it dynamically call toNumber if the `typeof num` it's passed is a string.

We have `satisfiesInputShape` which *might* be able to handle most of this, if it isn't already...? I need to confirm that `((join (0x) (08)) + (join (0x) (10))`, for example, *doesn't* work right now, because seemingly that's exactly what `satisfiesInputShape` [exists](https://github.com/leopard-js/sb-edit/blob/19bd48ef3e7747c2d5d430b37ec733dc95f8e42e/src/io/leopard/toLeopard.ts#L1814-L1816) to [handle](https://github.com/leopard-js/sb-edit/blob/19bd48ef3e7747c2d5d430b37ec733dc95f8e42e/src/io/leopard/toLeopard.ts#L1826-L1828). (This example should translate as `this.toNumber("0x" + "08") + this.toNumber("0x" + "10")`.)

**Comment by ZXMushroom63** [comment]
On the topic of `toNumber()` usage, I have found that Leopard sometimes uses it on strings?
<img alt="LeopardDetector" src="https://github.com/leopard-js/sb-edit/assets/116805577/e527a787-7230-45ba-9d9d-8edcd816c0c7" width="373.5" height="256.5">
Gets converted in to:
```javascript
  *whenGreenFlagClicked() {
    this.stage.vars.isleopard = 0;
    if (this.toNumber(this.letterOf("a", 0)) === 0) {
      this.stage.vars.isleopard = 1;
    }
  }
```
Anyway, this is the last Leopard-bug I've found while convert my project. Good luck fixing the lot! :)

**Comment by towerofnix** [comment]
That's another good catch! Since this is a fairly large issue, we're tracking it separately now: #123.

**Comment by towerofnix** [comment]
Things for me to verify, and split into multiple issues, if applicable:

* Comparing against an empty string has to work. Don't treat an empty string as zero for the equals block.
* When serialized to a number, hex inputs should be serialized in hex format: "0x10" -> `0x10`, not `16`.
* Unlike decimal inputs, hex inputs **should not** serialize to a number when used in an `InputShape.Any` input - importantly, this is the shape used in "set variable to" and "add to list", where it's crucial that the value be saved as a string.
  * This is for compatibility with e.g. "letter 2 of 0x10", which is "x", not "6".
  * Decimal values avoid issues with being treated as strings because JavaScript stringifies numbers in base 10 (by default).
  * There are probably other type issues with storing numbers in variables instead of strings, even in the decimal case, but those can be addressed separately...
* Confirm Leopard's `this.compare` is working fine with hexadecimal values (e.g. `(join "1" "6") = (join "0x" "10")`). I think it's OK but worth double checking.
